torch/testing/_internal/common_utils.py

```diff
@@ -1820,6 +1820,32 @@ def __init__(self, actual, expected, *, rtol_override=0.0, atol_override=0.0, **
         self.rtol = max(self.rtol, rtol_override)
         self.atol = max(self.atol, atol_override)
 
+        # This is a slow and ugly hack to allow the comparison of hybrid sparse CSR tensors with strided ones. If
+        # `check_layout=False` (default), the tensors will be converted to strided by calling `.to_dense()` on them.
+        # However, this is not yet supported for hybrid sparse CSR and thus we need to do it manually for now.
+        # FIXME: Remove this as soon as `.to_dense` is supported for hybrid sparse CSR tensors
+        if not self.check_layout:
+            self.actual, self.expected = self._handle_hybrid_sparse_csr(self.actual, self.expected)
+
+    def _handle_hybrid_sparse_csr(self, actual, expected):
+        if not ((actual.layout is torch.sparse_csr) ^ (expected.layout is torch.sparse_csr)):
+            return actual, expected
+        def to_dense(tensor):
+            if tensor.layout is not torch.sparse_csr:
+                return tensor
+
+            csr_dims = tensor.values().ndim
+
+            def partial_to_dense(tensor):
+                if tensor.ndim == csr_dims:
+                    return tensor.to_dense()
+
+                return torch.stack([partial_to_dense(sub_tensor) for sub_tensor in tensor])
```

Comment:
Notice that looping a CSR/CSC/BSR/BSC tensors like above will result in COO tensors at the first recursion level:
```python
In [14]: a = torch.sparse_csr_tensor([0, 2, 4], [0, 1, 0, 1], [[1, 11], [2, 12], [3, 13], [4, 14]])

In [15]: for item in a:
    ...:     print(item)
    ...: 
tensor(indices=tensor([[0, 1]]),
       values=tensor([[ 1, 11],
                      [ 2, 12]]),
       size=(2, 2), nnz=2, layout=torch.sparse_coo)
tensor(indices=tensor([[0, 1]]),
       values=tensor([[ 3, 13],
                      [ 4, 14]]),
       size=(2, 2), nnz=2, layout=torch.sparse_coo)
```
so that the next iterations can safely use `to_dense()`. However, to support also a batch of hybrid CSR/CSC/BSR/BSC tensors, like,
```python
a = torch.sparse_csr_tensor([[0, 2, 4], [0, 2, 4]], [[0, 1, 0, 1], [0, 1, 0, 1]], [[[1, 11], [2, 12], [3, 13], [4, 14]], [[10, 110], [20, 120], [30, 130], [40, 140]]])
```
some recursion is required but the stopping criteria should be based on whether the input tensor can handle `to_dense` or not:
```suggestion
            def partial_to_dense(tensor):
                if tensor.layout not in compressed_sparse_layouts or tensor.values().ndim == 1:
                    return tensor.to_dense()
                return torch.stack([partial_to_dense(sub_tensor) for sub_tensor in tensor])
```